### PR TITLE
Fix upload button on comment textareas

### LINKED ISSUE #1242: Upload icon missing
From https://github.com/sindresorhus/refined-github/issues/1179#issuecomment-378730057:

screenshot from 2018-04-04 21-51-54

Workaround: Ctrl + U.

### PR DESCRIPTION
**Pull request by notlmn**
Fixes #1242.

Explanation: https://github.com/sindresorhus/refined-github/issues/1242#issuecomment-381403293

-----------

The code looks a bit hacky, but works.

Open to any other suggestions or modifications!

### CODE REVIEW

File: source/features/upload-button.js [hunk 1]
```diff
@@ -3,22 +3,42 @@ import select from 'select-dom';
...
+					<button type="button" class="toolbar-item rgh-upload-btn">
+						{icons.cloudUpload()}
+					</button>
+				);
+
+				form.classList.add('rgh-has-upload-field');
+			}
+		}, {
+			childList: true,
+			subtree: true
```
Comment (bfred-it) [inline source/features/upload-button.js]: This config isn't required I think. `childList` should be enough and that's the default.

File: source/features/upload-button.js [hunk 2]
```diff
@@ -3,22 +3,42 @@ import select from 'select-dom';
...
-		}
-		toolbarPosition.after(
-			<button type="button" class="toolbar-item rgh-upload-btn">
-				{icons.cloudUpload()}
-			</button>
-		);
+
+		const toolbar = select('markdown-toolbar', form);
+
+		const observer = observeEl(toolbar, () => {
```
Comment (bfred-it) [inline source/features/upload-button.js]: The code inside this function should probably be what was before, it accomplishes the same in fewer lines:

```js

		form.classList.add('rgh-has-upload-field');
		const toolbarPosition = select('.js-saved-reply-container', form);
		if (!toolbarPosition) {
			continue;
		}
		toolbarPosition.after(
			<button type="button" class="toolbar-item rgh-upload-btn">
				{icons.cloudUpload()}
			</button>
		);
```

File: source/features/upload-button.js [hunk 3]
```diff
@@ -3,22 +3,42 @@ import select from 'select-dom';
...
+			if (form.classList.contains('rgh-has-upload-field')) {
+				observer.disconnect();
```
Comment (bfred-it) [inline source/features/upload-button.js]: This should also be called after the first run, like after line 37

File: source/features/upload-button.js [hunk 4]
```diff
@@ -38,6 +58,13 @@ function handleKeydown(event) {
...
 	addButtons();
-	delegate('.rgh-has-upload-field', 'keydown', handleKeydown);
+
+	if (delegations) {
+		delegations.map(d => d.destroy());
+		delegations = null;
+	}
+
+	delegations = delegate('.rgh-has-upload-field', 'keydown', handleKeydown);
+
```
[the remark this answers is not in the document]
Comment (notlmn) [reply source/features/upload-button.js]: Yes, `delegate` is global, but `ajaxedPagesHandler()` gets called for ajax loads (like moving from one issue to the other), and causes the listeners to be added again and again, causing mutliple *click events* on button (where the file selection dialog opens multiples times based on the number of event listeners added to it).

[the remark this answers is not in the document]
Comment (bfred-it) [reply source/features/upload-button.js]: Gotcha, good catch. What I've been doing for that is moving functions **out** of `ajaxedPagesHandler` into `domReady` or `init` and then, if necessary, use the ajax handler **inside** the feature.

In this case:

```js
import {safeOnAjaxedPages} from '../libs/utils';

...

safeOnAjaxedPages(addButtons);
```

File: source/features/upload-button.js (hunk 2 above)
Comment (notlmn) [reply source/features/upload-button.js]: Mutations can happen on multiple elements on the header element, we cannot assume that after the first mutation, the toolbar is ready.

So we keep checking, and only disconnect when the button is appended.

File: source/features/upload-button.js (hunk 3 above)
Comment (notlmn) [reply source/features/upload-button.js]: Answered [above](https://github.com/sindresorhus/refined-github/pull/1262#discussion_r181586751).

File: source/features/upload-button.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (notlmn) [reply source/features/upload-button.js]: That would be great handling ajaxed pages for individual feature!

But wouldn't it be good to do all the refactor in a separate PR? (I'm up for this!)

[the remark this answers is not in the document]
Comment (bfred-it) [reply source/features/upload-button.js]: Done in cf2f821

File: source/features/upload-button.js (hunk 2 above)
Comment (bfred-it) [reply source/features/upload-button.js]: True, I kept the check, but shortened the function in f688b2

**Comment by notlmn** [comment]
Works as intended!

**Pull request merged**